README.md
# configd code design

## Overview

The configd package is written in Go.  It provides functionality to parse,
compile and validate Yang schemas, and to validate base configurations and
configuration updates against those schemas.  Further to this it manages the
configuration change process (set/delete, commit etc).

Much of the subsidiary functionality, eg the YANG compiler and handling for
configd YANG extensions, is implemented in separate repositories.

## Packages in this repository

The following packages are included in this repository.

### Yang cmds

  * yang2path
  * yang2rev
  * yangc       - standalone version of YANG compiler (inc cfgd xtns)
  * yanggraph

### Configd cmds

  * args2cpath
  * callrpc
  * cfgcli      - implements 'config' mode commands
  * cfgdiff
  * cfgparse
  * cfgread
  * configd     - configd process (parses YANG, config, runs sessions)
  * configdcaps
  * featcaps
  * gettree
  * normalize

### Configd packages

  * configd (configd)         - authentication, logging, error formatting
  * commit  (configd/commit)  - builds config trees for commit
  * session (configd/session) - session management
  * server  (configd/server)  - server for incoming RPCs


debian/changelog
configd (2.12.2) unstable; urgency=medium

  * Fix over-optimised leafref allowed values cacheing.

 -- William Ivory <william.ivory@intl.att.com>  Tue, 28 Sep 2021 16:58:25 +0100

configd (2.12.1) unstable; urgency=medium

  [ Paul Carson ]
  * Block commit when disk is read-only (Fixes: VRVDR-55566)

 -- William Ivory <william.ivory@intl.att.com>  Tue, 31 Aug 2021 13:00:08 +0100

configd (2.12) unstable; urgency=medium

  [ Nicholas Brown ]
  * dh-systemd is part of debhelper

  [ William Ivory ]
  * Bullseye build
  * Move configd flag.Parse() from init() to main()

  [ Paul Carson ]
  * Inhibit log message: Login user id not set for pid (Fixes: VRVDR-50782)
  * CopyConfig: Support selectable config encoding
  * Support ValidateConfig
  * Unit tests to verify validate config (Closes: VRVDR-52942)

  [ William Ivory ]
  * Move logic for getting component state to modelset_extension method.
  * Switch DBUS calls to VCI, using ComponentManager.
  * Update copyrights

  [ Paul Carson ]
  * YANG: Support Yang choices (Closes: VRVDR-6224)

  [ William Ivory ]
  * Update tests for modified XChildren() API.
  * Switch getfulltree log message from debug to error.

  [ John Southworth ]
  * Call all RPCs via configd

  [ William Ivory ]
  * Add tests for leafref performance.
  * Use schema.NewCompMgr()
  * Move Set/GetRunning and GetState from ms to compMgr
  * Move componentMappings to component manager

 -- Paul Carson <paul.carson@intl.att.com>  Thu, 01 Jul 2021 13:53:02 +0100

configd (2.11) unstable; urgency=medium

  [ Duncan Eastoe ]
  * configd.tmpfile: Make /var/tmp/configd dir
  * server: use temporary directory for load/save ops
  * config_mgmt: remove spawnCommandAsCallerFn
  * config_mgmt: split spawnCommandAsCaller()
  * server: Add userProcessReader
  * load_keys: don't load complete file into memory
  * load_keys: update session as keys are parsed
  * server: Switch load/loadkey to userProcessReader
  * config_mgmt: os.Chown() --> File.Chown()
  * config_mgmt: copyFile & uploadFile take *os.File
  * config_mgmt.go: Check for os.File.Sync() error
  * session: Add concept of session "owner"
  * server: mark "superuser" connections in Context
  * session: SessionMgr.Get() accepts Context param
  * session: Add initial session manager UTs
  * session: Add SessionMgr.lookup()
  * session: Restrict access to non-shared sessions
  * session: Allow sessions to be shared or un-shared
  * session: Allow superuser to get any session
  * configd: Add SetupSessionShared RPC

  [ William Ivory ]
  * Updated InfoTags handling for test mgmt errors
  * Add CopyConfig API
  * Add AAA authorization wrapper to CopyConfig.
  * Return error list from load.go merge_tree()
  * Fix crash in expandPath()

 -- William Ivory <william.ivory@intl.att.com>  Tue, 23 Feb 2021 13:43:39 +0000

configd (2.10) unstable; urgency=medium

  [ Nicholas Brown ]
  * Reuse the same build worker node for the docker container

  [ John Southworth ]
  * Tweak memory settings in the configd process

  [ Paul Carson ]
  * Configure top level defaults after boot

  [ Duncan Eastoe ]
  * edit_config: default-operation none may merge cfg

  [ William Ivory ]
  * Encode MgmtErrors explicitly; decode in client.go

  [ Duncan Eastoe ]
  * edit_config_auth_test.go: Fatalf() --> Errorf()
  * edit_config.go: Skip accounting on test only
  * Update to use latest NewTestAutherCommandRequest
  * aaa.go: Move AAA related funcs from dispatcher.go
  * dispatcher: split accounted RPC handlers
  * dispatcher: Migrate to auth.TaskAccounter API
  * edit_config: Migrate to auth.TaskAccounter API
  * cli_format.go: Avoid "[] is not valid" in RPC err

 -- William Ivory <william.ivory@intl.att.com>  Fri, 25 Sep 2020 08:45:06 +0100

configd (2.9) unstable; urgency=medium

  [ John Southworth ]
  * Create a binary argument encoder to speed up argument encoding.

  [ Nicholas Brown ]
  * Specify SyslogIdentifier= in the systemd service file (Fixes: VRVDR-50783)

  [ John Southworth ]
  * Create rfc7951 merge helper that understands our schema representation
  * Add new library to debian/copyright
  * Add additional Connect function.
  * Fix SPDX reference for configd/client/vci.go
  * Use new shell library for quoting in cfgcmds

  [ Nicholas Brown ]
  * Add a revive.toml file with exported disabled

 -- William Ivory <william.ivory@intl.att.com>  Wed, 22 Jul 2020 11:35:13 +0100

configd (2.8) unstable; urgency=medium

  [ Erik Ziegenbalg ]
  * Configd.postinstall: Tab noise cleanup
  * Configd.postinstall: Sanatize triggered)

  [ Paul Carson ]
  * Support NETCONF confirmed commit (Closes: VRVDR-32867)

  [ Duncan Eastoe ]
  * config_mgmt: Update user sandbox path
  * configd.tmpfile: /var/run -> /run

  [ Paul Carson ]
  * Prevent undesired errors on confirming commit (Fixes: VRVDR-50002)

  [ William Ivory ]
  * Add must threshold debug to cfgdbg utility
  * Pass must debug threshold to validation code

  [ Paul Carson ]
  * Improve cancel-commit command (Fixes: VRVDR-50028)

  [ William Ivory ]
  * Fix escaping for backticks
  * Fix auto-completion of '$$'

  [ Paul Carson ]
  * Prevent scheulling revert if commit fails (Fixes: VRVDR-50125)
  * Improve confirmed-commit logging (Closes: VRVDR-50125)

 -- William Ivory <william.ivory@intl.att.com>  Fri, 03 Apr 2020 11:35:57 +0100

configd (2.7) unstable; urgency=medium

  * DANOS Import Master

 -- Vyatta Package Maintainers <DL-vyatta-help@att.com>  Fri, 15 Nov 2019 11:37:35 +0000

configd (2.6.3.1) unstable; urgency=medium

  * DANOS Import

 -- Vyatta Package Maintainers <DL-vyatta-help@att.com>  Tue, 12 Nov 2019 09:14:46 +0000


cmd/yangc/yangc.go
// Copyright (c) 2018-2019, AT&T Intellectual Property. All rights reserved.
//
// Copyright (c) 2014-2017 by Brocade Communications Systems, Inc.
// All rights reserved.
//
// SPDX-License-Identifier: LGPL-2.1-only

package main

import (
	"flag"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/danos/config/platform"
	"github.com/danos/config/schema"
	"github.com/danos/yang/compile"
	"github.com/danos/yang/xpath"
	"github.com/danos/yang/xpath/xutils"
	"github.com/go-ini/ini"
)

var skip bool
var showNPContainerWarnings bool
var capabilities string
var fullSchema bool
var customXpathFunctions string
var resourcesDir string
var testPlatform string

func usage() {
	_, file := filepath.Split(os.Args[0])
	fmt.Fprintf(os.Stderr, "Usage of %s [flags] [<yang-dir>]:\n\n", file)

	flag.PrintDefaults()

	usageInfo := `
  This utility parses YANG files, checking for any errors.  It has 2 different
  sets of mutually-exclusive parameters for this:

  (a) Simple YANG and feature validation

      Use '-capabilities' (features) and <yang-dir>

  (b) Specific platform validation using 'resources' directory created by
      DRAM or get_iso_yang utilities.

      Use '-resources-dir' (rqd) and '-platform' (optional).  If no platform
      is specified, the 'base' platform will be used.

`
	fmt.Fprintf(os.Stderr, usageInfo)

}

func init() {
	flag.BoolVar(&skip, "i", false, "Ignore unknown symbols")
	flag.StringVar(&capabilities, "capabilities", "",
		"File specifying system capabilities/features")
	flag.BoolVar(&fullSchema, "full", true, "Parse full schema (inc state)")
	flag.BoolVar(&showNPContainerWarnings, "show-np-cont-warnings", false,
		"Show warnings relating to must statements on NP containers")
	flag.StringVar(&customXpathFunctions, "custom-xpath-functions", "",
		"Additional custom XPATH functions")
	flag.StringVar(&resourcesDir, "resources-directory", "",
		"Resources Directory (eg for DRAM)")
	flag.StringVar(&testPlatform, "platform", "",
		"Specific platform to test (default is BASE platform)")
}

const (
	featuresSubdir           = "features"
	vyattaPlatformSubdir     = "vyatta-platform"
	vyattaPlatformYangSubdir = vyattaPlatformSubdir + "/yang"
	xpathPluginDir           = "/lib/xpath/plugins"
)

// getCustomFunctionList - merge functions from INI files with user's list
func getValidCustomFunctionList(customXpathFunctions string) []string {

	functions := getFunctionsFromIniFiles()

	for _, function := range strings.Split(customXpathFunctions, ",") {
		if function != "" {
			functions = append(functions, function)
		}
	}

	return functions
}

func getFunctionsFromIniFiles() []string {
	var functions []string

	files, _ := filepath.Glob(filepath.Join(xpathPluginDir, "*.ini"))
	for _, file := range files {
		iniFile, err := ini.Load(file)
		if err != nil {
			// Simply ignore problematic files.  Best effort here.
			continue
		}
		for _, section := range iniFile.Sections() {
			functions = append(functions, section.Name())
		}
	}

	return functions
}

// getPlatform returns the relevant platform definition
//
// Base platform (empty definition) is returned if the platform name is
// the empty string.
func getPlatform(dir, platName string) (*platform.Definition, error) {

	// Either YangDir specified, or resources with no platform specified.
	// Return the 'base' platform definition in this case.
	if platName == "" {
		return &platform.Definition{}, nil
	}

	// Get platform definitions
	platDir := filepath.Join(dir, vyattaPlatformSubdir)
	plats := platform.NewPlatform().
		PlatformBaseDir(platDir).
		LoadDefinitions()

	// Find specific platform
	plat, ok := plats.Platforms[platName]
	if !ok {
		return nil, fmt.Errorf("Unable to find definition for platform '%s'",
			platName)
	}
	return plat, nil

}

func createYangLocator(
	args []string,
	resources string,
	plat *platform.Definition,
) compile.YangLocator {

	if len(args) > 0 {
		return compile.YangDirs(args[0])
	}

	files := make([]string, 0)
	for _, file := range plat.Yang {
		yangFile := filepath.Join(resources, vyattaPlatformYangSubdir, file)
		files = append(files, yangFile)
	}
	return compile.YangLocations(
		compile.YangDirs(filepath.Join(resources, "yang")),
		compile.YangFiles(files...))
}

func createFeatureChecker(
	caps,
	resources string,
	plat *platform.Definition,
) compile.FeaturesChecker {

	// In basic case, we have a single directory of features, specified by
	// <caps>
	if caps != "" {
		return compile.FeaturesFromLocations(true, caps)
	}

	// If we have a resources directory, features consist of:
	//   <resources>/features
	//     - default enabled features
	//     - user-defined extra enabled features
	//   platform-specific features (enabled and disabled)
	return compile.MultiFeatureCheckers(
		compile.FeaturesFromLocations(
			true, filepath.Join(resourcesDir, featuresSubdir)),
		compile.FeaturesFromNames(true, plat.Features...),
		compile.FeaturesFromNames(false, plat.DisabledFeatures...))
}

// validateArgs - check params are valid, mutexes honoured ...
//
// 	 YangDir  caps  resources platform | Notes
// 1    n    	n	    n        n     | Invalid (no YANG specified)
// 2    n    	n	    n        y     | Invalid (platform w/o resources)
// 3    n    	n	    y        n     | Base Platform with resources
// 4    n    	n	    y        y     | Platform with resources
// 5    n    	y	    n        n     | Invalid (no YangDir specified)
// 6    n    	y	    n        y     | Invalid (platform w/o resources)
// 7    n    	y	    y        n     | Invalid (caps + resources not allowed)
// 8    n    	y	    y        y     | Invalid (caps + resources not allowed)
// 9    y    	n	    n        n     | Base Platform with YangDir
// 10   y    	n	    n        y     | Invalid (plat not valid with YangDir)
// 11   y    	n	    y        n     | Invalid (two sources of YANG)
// 12   y    	n	    y        y     | Invalid (two sources of YANG)
// 13   y    	y	    n        n     | Base Platform with YangDir
// 14   y    	y	    n        y     | Invalid (plat w/o resources)
// 15   y    	y	    y        n     | Invalid (two sources of YANG)
// 16   y    	y	    y        y     | Invalid (two sources of YANG)
func validateArgs(args []string) bool {

	// Exactly one source of YANG
	if (len(args) == 0) == (resourcesDir == "") {
		return false
	}

	// Capabilities only with YangDir not Resources
	if capabilities != "" && resourcesDir != "" {
		return false
	}

	// Platform can only be non-base with Resources
	if testPlatform != "" && resourcesDir == "" {
		return false
	}

	return true
}

func createUserFnChecker(
	customFns string,
) func(name string) (*xpath.Symbol, bool) {

	return func(name string) (*xpath.Symbol, bool) {
		allowedFns := getValidCustomFunctionList(customFns)
		for _, fn := range allowedFns {
			if fn == name {
				return xpath.NewDummyFnSym(name), true
			}
		}
		return nil, false
	}
}

func main() {
	flag.Usage = usage
	flag.Parse()
	args := flag.Args()

	if !validateArgs(args) {
		usage()
		os.Exit(1)
	}

	plat, err := getPlatform(resourcesDir, testPlatform)
	if err != nil {
		fmt.Fprintf(os.Stderr, "%s", err)
		os.Exit(1)
	}

	yangLocator := createYangLocator(args, resourcesDir, plat)

	featChecker := createFeatureChecker(capabilities, resourcesDir, plat)

	userFnChecker := createUserFnChecker(customXpathFunctions)

	_, warns, err := schema.CompileDirWithWarnings(
		&compile.Config{
			YangLocations: yangLocator,
			Features:      featChecker,
			SkipUnknown:   skip,
			Filter:        compile.IsConfigOrState(),
			UserFnCheckFn: userFnChecker},
		nil)

	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}

	if !showNPContainerWarnings {
		warns = xutils.RemoveNPContainerWarnings(warns)
	}

	if len(warns) != 0 {
		fmt.Fprintf(os.Stdout, "\nXPATH path validation warnings:\n\n")
		for _, warn := range warns {
			fmt.Fprintf(os.Stdout, "%s\n----\n", warn)
		}
		os.Exit(1)
	}
	os.Exit(0)
}
